File: TextosJapones/src/main/java/org/jisho/textosJapones/controller/mangas/MangasProcessarController.java
package org.jisho.textosJapones.controller.mangas;

import java.net.URL;
import java.util.List;
import java.util.ResourceBundle;

import org.jisho.textosJapones.Run;
import org.jisho.textosJapones.components.CheckBoxTreeTableCellCustom;
import org.jisho.textosJapones.components.notification.AlertasPopup;
import org.jisho.textosJapones.controller.GrupoBarraProgressoController;
import org.jisho.textosJapones.controller.MenuPrincipalController;
import org.jisho.textosJapones.model.entities.Manga;
import org.jisho.textosJapones.model.entities.MangaCapitulo;
import org.jisho.textosJapones.model.entities.MangaPagina;
import org.jisho.textosJapones.model.entities.MangaTabela;
import org.jisho.textosJapones.model.entities.MangaVolume;
import org.jisho.textosJapones.model.enums.Language;
import org.jisho.textosJapones.model.exceptions.ExcessaoBd;
import org.jisho.textosJapones.model.services.MangaServices;
import org.jisho.textosJapones.processar.ProcessarMangas;

import com.jfoenix.controls.JFXAutoCompletePopup;
import com.jfoenix.controls.JFXButton;
import com.jfoenix.controls.JFXCheckBox;
import com.jfoenix.controls.JFXComboBox;
import com.jfoenix.controls.JFXTextField;
import com.nativejavafx.taskbar.TaskbarProgressbar;
import com.nativejavafx.taskbar.TaskbarProgressbar.Type;

import javafx.application.Platform;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.value.ChangeListener;
import javafx.beans.value.ObservableValue;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.concurrent.Task;
import javafx.event.EventHandler;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.geometry.Pos;
import javafx.scene.control.ProgressBar;
import javafx.scene.control.TreeItem;
import javafx.scene.control.TreeTableCell;
import javafx.scene.control.TreeTableColumn;
import javafx.scene.control.TreeTableView;
import javafx.scene.control.cell.TreeItemPropertyValueFactory;
import javafx.scene.input.KeyCode;
import javafx.scene.input.KeyEvent;
import javafx.scene.layout.AnchorPane;
import javafx.scene.robot.Robot;
import javafx.util.Callback;

public class MangasProcessarController implements Initializable {

	@FXML
	protected AnchorPane apRoot;

	@FXML
	private JFXButton btnCarregar;

	@FXML
	private JFXButton btnProcessar;

	@FXML
	private JFXComboBox<String> cbBase;

	@FXML
	private JFXTextField txtManga;

	@FXML
	private JFXCheckBox ckbProcessados;

	@FXML
	private JFXCheckBox ckbMarcarTodos;

	@FXML
	private TreeTableView<Manga> treeBases;

	@FXML
	private TreeTableColumn<Manga, Boolean> treecMacado;

	@FXML
	private TreeTableColumn<Manga, String> treecBase;

	@FXML
	private TreeTableColumn<Manga, String> treecManga;

	@FXML
	private TreeTableColumn<Manga, Integer> treecVolume;

	@FXML
	private TreeTableColumn<Manga, Float> treecCapitulo;

	@FXML
	private TreeTableColumn<Manga, Integer> treecPagina;

	@FXML
	private TreeTableColumn<Manga, String> treecNomePagina;

	@FXML
	private JFXButton btnTransferir;

	@FXML
	private JFXTextField txtBaseOrigem;

	@FXML
	private JFXTextField txtBaseDestino;

	@FXML
	private JFXCheckBox ckbCriarBase;

	@FXML
	private JFXComboBox<Language> cbLinguagem;

	@FXML
	private ProgressBar barraProgressoVolumes;

	@FXML
	private ProgressBar barraProgressoCapitulos;

	@FXML
	private ProgressBar barraProgressoPaginas;

	private ProcessarMangas mangas;
	private MangaServices service = new MangaServices();
	private ObservableList<MangaTabela> TABELAS;

	private MangasController controller;

	public void setControllerPai(MangasController controller) {
		this.controller = controller;
	}

	public MangasController getControllerPai() {
		return controller;
	}

	@FXML
	private void onBtnProcessar() {
		if (btnProcessar.getAccessibleText().equalsIgnoreCase("PROCESSANDO") && mangas != null) {
			mangas.setDesativar(true);
			return;
		}

		btnProcessar.setAccessibleText("PROCESSANDO");
		btnProcessar.setText("Pausar");
		btnCarregar.setDisable(true);

		if (mangas == null)
			mangas = new ProcessarMangas(this);

		treeBases.setDisable(true);
		MenuPrincipalController.getController().getLblLog().setText("Iniciando o processamento dos mangas...");

		switch (cbLinguagem.getSelectionModel().getSelectedItem()) {
			case ENGLISH:
				mangas.processarTabelasIngles(TABELAS);
				break;
			case JAPANESE:
				mangas.processarTabelasJapones(TABELAS);
				break;
			default:
		}
	}

	@FXML
	private void onBtnCarregar() {
		carregar();
	}

	@FXML
	private void onBtnMarcarTodos() {
		marcarTodosFilhos(treeBases.getRoot(), ckbMarcarTodos.isSelected());
		treeBases.refresh();
	}

	@FXML
	private void onBtnTransferir() {
		transferir();
	}

	public void habilitar() {
		treeBases.setDisable(false);
		MenuPrincipalController.getController().getLblLog().setText("");
		btnProcessar.setAccessibleText("PROCESSAR");
		btnProcessar.setText("Processar");
		btnCarregar.setDisable(false);
		TaskbarProgressbar.stopProgress(Run.getPrimaryStage());
		getBarraProgressoVolumes().setProgress(0);
		getBarraProgressoCapitulos().setProgress(0);
		getBarraProgressoPaginas().setProgress(0);
	}

	public AnchorPane getRoot() {
		return apRoot;
	}

	public ProgressBar getBarraProgressoVolumes() {
		return barraProgressoVolumes;
	}

	public ProgressBar getBarraProgressoCapitulos() {
		return barraProgressoCapitulos;
	}

	public ProgressBar getBarraProgressoPaginas() {
		return barraProgressoPaginas;
	}

	private String BASE_ORIGEM, BASE_DESTINO;

	private Integer I;
	private String error;

	private void transferir() {
		if (txtBaseOrigem.getText().trim().equalsIgnoreCase(txtBaseDestino.getText().trim())) {
			AlertasPopup.AvisoModal(controller.getStackPane(), controller.getRoot(), null, "Aviso",
					"Favor informar outra base de destino.");
			return;
		}

		GrupoBarraProgressoController progress = MenuPrincipalController.getController().criaBarraProgresso();
		progress.getTitulo().setText("Mangas - Transferencia");
		progress.getLog().setText("Transferindo dados....");
		btnTransferir.setDisable(true);

		BASE_ORIGEM = txtBaseOrigem.getText().trim();
		BASE_DESTINO = txtBaseDestino.getText().trim();

		if (TaskbarProgressbar.isSupported())
			TaskbarProgressbar.showIndeterminateProgress(Run.getPrimaryStage());

		Task<Void> transferir = new Task<Void>() {

			@Override
			protected Void call() throws Exception {
				try {
					error = "";
					updateMessage("Carregando dados....");
					List<MangaVolume> lista = service.selectDadosTransferir(BASE_ORIGEM);

					if (ckbCriarBase.isSelected()) {
						updateMessage("Criando a base....");
						service.createDataBase(BASE_DESTINO);
					}

					updateMessage("Transferindo dados....");
					I = 0;
					for (MangaVolume volume : lista) {
						updateMessage("Transferindo dados.... " + volume.getManga());
						I++;
						service.insertDadosTransferir(BASE_DESTINO, volume);
						updateProgress(I, lista.size());

						Platform.runLater(() -> {
							if (TaskbarProgressbar.isSupported())
								TaskbarProgressbar.showCustomProgress(Run.getPrimaryStage(), I, lista.size(),
										Type.NORMAL);
						});
					}

				} catch (ExcessaoBd e) {
					e.printStackTrace();
					error = e.getMessage();
				}
				return null;
			}

			@Override
			protected void succeeded() {
				Platform.runLater(() -> {
					progress.getBarraProgresso().progressProperty().unbind();
					progress.getLog().textProperty().unbind();
					MenuPrincipalController.getController().destroiBarraProgresso(progress, "");

					btnTransferir.setDisable(false);
					getBarraProgressoVolumes().setProgress(0);
					TaskbarProgressbar.stopProgress(Run.getPrimaryStage());

					if (!error.isEmpty())
						AlertasPopup.ErroModal(controller.getStackPane(), controller.getRoot(), null, "Erro", error);
					else
						AlertasPopup.AvisoModal(controller.getStackPane(), controller.getRoot(), null, "Aviso",
								"Transferência concluida.");
				});

			}
		};
		progress.getBarraProgresso().progressProperty().bind(transferir.progressProperty());
		progress.getLog().textProperty().bind(transferir.messageProperty());
		Thread t = new Thread(transferir);
		t.start();
	}

	private Boolean PROCESSADOS;
	private String BASE;
	private String MANGA;
	private Language LINGUAGEM;
	private TreeItem<Manga> DADOS;

	private void carregar() {
		MenuPrincipalController.getController().getLblLog().setText("Carregando dados dos mangas...");
		btnCarregar.setDisable(true);
		btnProcessar.setDisable(true);
		treeBases.setDisable(true);

		PROCESSADOS = ckbProcessados.isSelected();
		BASE = cbBase.getValue() != null ? cbBase.getValue().trim() : "";
		MANGA = txtManga.getText().trim();
		LINGUAGEM = cbLinguagem.getSelectionModel().getSelectedItem();

		getBarraProgressoCapitulos().setProgress(-1);
		getBarraProgressoVolumes().setProgress(-1);

		if (TaskbarProgressbar.isSupported())
			TaskbarProgressbar.showIndeterminateProgress(Run.getPrimaryStage());

		// Criacao da thread para que esteja validando a conexao e nao trave a tela.
		Task<Void> carregaItens = new Task<Void>() {

			@Override
			protected Void call() throws Exception {
				try {
					service = new MangaServices();
					TABELAS = FXCollections.observableArrayList(service.selectTabelas(!PROCESSADOS, false, BASE, LINGUAGEM, MANGA));
					DADOS = getTreeData();
				} catch (ExcessaoBd e) {
					e.printStackTrace();
				}
				return null;
			}

			@Override
			protected void succeeded() {
				Platform.runLater(() -> {
					treeBases.setRoot(DADOS);
					MenuPrincipalController.getController().getLblLog().setText("");
					ckbMarcarTodos.setSelected(true);
					btnCarregar.setDisable(false);
					btnProcessar.setDisable(false);
					treeBases.setDisable(false);
					getBarraProgressoCapitulos().setProgress(0);
					getBarraProgressoVolumes().setProgress(0);
					TaskbarProgressbar.stopProgress(Run.getPrimaryStage());
				});

			}
		};
		Thread t = new Thread(carregaItens);
		t.start();
	}

	private TreeItem<Manga> getTreeData() {
		TreeItem<Manga> itmRoot = new TreeItem<Manga>(new Manga("...", ""));
		for (MangaTabela tabela : TABELAS) {
			tabela.setManga("...");
			TreeItem<Manga> itmTabela = new TreeItem<Manga>(tabela);
			TreeItem<Manga> itmManga = null;
			String volumeAnterior = "";
			for (MangaVolume volume : tabela.getVolumes()) {
				// Implementa um nivel por tipo
				if (!volume.getManga().equalsIgnoreCase(volumeAnterior) || itmManga == null) {
					volumeAnterior = volume.getManga();
					volume.setBase(tabela.getBase());
					itmManga = new TreeItem<Manga>(new Manga(tabela.getBase(), volume.getManga(), "..."));
					itmTabela.getChildren().add(itmManga);
					itmTabela.setExpanded(true);
				}

				volume.setBase(tabela.getBase());
				TreeItem<Manga> itmVolume = new TreeItem<Manga>(volume);

				for (MangaCapitulo capitulo : volume.getCapitulos()) {
					capitulo.setBase(tabela.getBase());
					capitulo.setNomePagina("...");
					TreeItem<Manga> itmCapitulo = new TreeItem<Manga>(capitulo);
					for (MangaPagina pagina : capitulo.getPaginas()) {
						pagina.addOutrasInformacoes(tabela.getBase(), volume.getManga(), volume.getVolume(),
								capitulo.getCapitulo(), volume.getLingua());
						itmCapitulo.getChildren().add(new TreeItem<Manga>(pagina));
					}
					itmVolume.getChildren().add(itmCapitulo);
				}

				itmManga.getChildren().add(itmVolume);
			}
			itmRoot.getChildren().add(itmTabela);
			itmRoot.setExpanded(true);
		}
		return itmRoot;
	}

	private void marcarTodosFilhos(TreeItem<Manga> treeItem, Boolean newValue) {
		treeItem.getValue().setProcessar(newValue);
		treeItem.getChildren().forEach(treeItemNivel2 -> marcarTodosFilhos(treeItemNivel2, newValue));
	}

	private void ativaTodosPai(TreeItem<Manga> treeItem, Boolean newValue) {
		if (treeItem.getParent() != null) {
			treeItem.getParent().getValue().setProcessar(newValue);
			ativaTodosPai(treeItem.getParent(), newValue);
		}
	}

	private void editaColunas() {
		// ==== (CHECK-BOX) ===
		treecMacado.setCellValueFactory(
				new Callback<TreeTableColumn.CellDataFeatures<Manga, Boolean>, ObservableValue<Boolean>>() {

					@Override
					public ObservableValue<Boolean> call(TreeTableColumn.CellDataFeatures<Manga, Boolean> param) {
						TreeItem<Manga> treeItem = param.getValue();
						Manga item = treeItem.getValue();
						SimpleBooleanProperty booleanProp = new SimpleBooleanProperty(item.isProcessar());

						booleanProp.addListener(new ChangeListener<Boolean>() {
							@Override
							public void changed(ObservableValue<? extends Boolean> observable, Boolean oldValue,
									Boolean newValue) {
								item.setProcessar(newValue);
								marcarTodosFilhos(treeItem, newValue);
								if (newValue) // Somente ativa caso seja true, pois ao menos um nó precisa estar ativo
									ativaTodosPai(treeItem, newValue);

								treeBases.refresh();
							}
						});

						return booleanProp;
					}
				});

		treecMacado.setCellFactory(new Callback<TreeTableColumn<Manga, Boolean>, TreeTableCell<Manga, Boolean>>() {
			@Override
			public TreeTableCell<Manga, Boolean> call(TreeTableColumn<Manga, Boolean> p) {
				CheckBoxTreeTableCellCustom<Manga, Boolean> cell = new CheckBoxTreeTableCellCustom<Manga, Boolean>();
				cell.setAlignment(Pos.CENTER);
				return cell;
			}
		});

	}

	private void linkaCelulas() {
		treecMacado.setCellValueFactory(new TreeItemPropertyValueFactory<Manga, Boolean>("processar"));
		treecBase.setCellValueFactory(new TreeItemPropertyValueFactory<>("base"));
		treecManga.setCellValueFactory(new TreeItemPropertyValueFactory<>("manga"));
		treecVolume.setCellValueFactory(new TreeItemPropertyValueFactory<>("volume"));
		treecCapitulo.setCellValueFactory(new TreeItemPropertyValueFactory<>("capitulo"));
		treecPagina.setCellValueFactory(new TreeItemPropertyValueFactory<>("pagina"));
		treecNomePagina.setCellValueFactory(new TreeItemPropertyValueFactory<>("nomePagina"));
		treeBases.setShowRoot(false);

		editaColunas();
	}

	private Robot robot = new Robot();

	public void initialize(URL arg0, ResourceBundle arg1) {
		cbLinguagem.getItems().addAll(Language.JAPANESE, Language.ENGLISH);
		cbLinguagem.getSelectionModel().selectFirst();
		
		try {
			cbBase.getItems().setAll(service.getTabelas());
		} catch (ExcessaoBd e) {
			e.printStackTrace();
			AlertasPopup.ErroModal("Erro ao carregar as tabelas", e.getMessage());
		}
		
		JFXAutoCompletePopup<String> autoCompletePopup = new JFXAutoCompletePopup<>();
		autoCompletePopup.getSuggestions().addAll(cbBase.getItems());

		autoCompletePopup.setSelectionHandler(event -> {
			cbBase.setValue(event.getObject());
		});

		cbBase.getEditor().textProperty().addListener(observable -> {
			autoCompletePopup.filter(item -> item.toLowerCase().contains(cbBase.getEditor().getText().toLowerCase()));
			if (autoCompletePopup.getFilteredSuggestions().isEmpty() || cbBase.showingProperty().get()
					|| cbBase.getEditor().getText().isEmpty())
				autoCompletePopup.hide();
			else
				autoCompletePopup.show(cbBase.getEditor());
		});

		cbBase.setOnKeyPressed(new EventHandler<KeyEvent>() {
			@Override
			public void handle(KeyEvent ke) {
				if (ke.getCode().equals(KeyCode.ENTER))
					robot.keyPress(KeyCode.TAB);
			}
		});

		txtManga.setOnKeyPressed(new EventHandler<KeyEvent>() {
			@Override
			public void handle(KeyEvent ke) {
				if (ke.getCode().equals(KeyCode.ENTER))
					robot.keyPress(KeyCode.TAB);
			}
		});

		linkaCelulas();
	}

	public static URL getFxmlLocate() {
		return MangasProcessarController.class.getResource("/view/mangas/MangaProcessar.fxml");
	}
}

File: ProcessaTexto/src/main/kotlin/br/com/fenix/processatexto/controller/mangas/MangasProcessarController.kt
package br.com.fenix.processatexto.controller.mangas

import br.com.fenix.processatexto.Run
import br.com.fenix.processatexto.components.CheckBoxTreeTableCellCustom
import br.com.fenix.processatexto.components.notification.AlertasPopup
import br.com.fenix.processatexto.controller.MenuPrincipalController
import br.com.fenix.processatexto.model.entities.Manga
import br.com.fenix.processatexto.model.entities.mangaextractor.MangaTabela
import br.com.fenix.processatexto.model.entities.mangaextractor.MangaVolume
import br.com.fenix.processatexto.model.enums.Language
import br.com.fenix.processatexto.processar.ProcessarMangas
import br.com.fenix.processatexto.service.MangaServices
import com.jfoenix.controls.*
import com.nativejavafx.taskbar.TaskbarProgressbar
import com.nativejavafx.taskbar.TaskbarProgressbar.Type
import javafx.application.Platform
import javafx.beans.property.SimpleBooleanProperty
import javafx.collections.FXCollections
import javafx.collections.ObservableList
import javafx.concurrent.Task
import javafx.fxml.FXML
import javafx.fxml.Initializable
import javafx.geometry.Pos
import javafx.scene.control.*
import javafx.scene.control.cell.TreeItemPropertyValueFactory
import javafx.scene.input.KeyCode
import javafx.scene.layout.AnchorPane
import javafx.scene.robot.Robot
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.net.URL
import java.sql.SQLException
import java.util.*


class MangasProcessarController : Initializable {

    private val LOGGER: Logger = LoggerFactory.getLogger(MangasProcessarController::class.java)

    @FXML
    private lateinit var apRoot: AnchorPane

    @FXML
    private lateinit var btnCarregar: JFXButton

    @FXML
    private lateinit var btnProcessar: JFXButton

    @FXML
    private lateinit var cbBase: JFXComboBox<String>

    @FXML
    private lateinit var txtManga: JFXTextField

    @FXML
    private lateinit var ckbProcessados: JFXCheckBox

    @FXML
    private lateinit var ckbMarcarTodos: JFXCheckBox

    @FXML
    private lateinit var treeBases: TreeTableView<Manga>

    @FXML
    private lateinit var treecMacado: TreeTableColumn<Manga, Boolean>

    @FXML
    private lateinit var treecBase: TreeTableColumn<Manga, String>

    @FXML
    private lateinit var treecManga: TreeTableColumn<Manga, String>

    @FXML
    private lateinit var treecVolume: TreeTableColumn<Manga, Int>

    @FXML
    private lateinit var treecCapitulo: TreeTableColumn<Manga, Float>

    @FXML
    private lateinit var treecPagina: TreeTableColumn<Manga, Int>

    @FXML
    private lateinit var treecNomePagina: TreeTableColumn<Manga, String>

    @FXML
    private lateinit var btnTransferir: JFXButton

    @FXML
    private lateinit var txtBaseOrigem: JFXTextField

    @FXML
    private lateinit var txtBaseDestino: JFXTextField

    @FXML
    private lateinit var ckbCriarBase: JFXCheckBox

    @FXML
    private lateinit var cbLinguagem: JFXComboBox<Language>

    @FXML
    lateinit var barraProgressoVolumes: ProgressBar

    @FXML
    lateinit var barraProgressoCapitulos: ProgressBar

    @FXML
    lateinit var barraProgressoPaginas: ProgressBar

    private var mangas: ProcessarMangas? = null
    private var service: MangaServices = MangaServices()
    private var TABELAS: ObservableList<MangaTabela> = FXCollections.observableArrayList()

    private lateinit var controller: MangasController
    var controllerPai: MangasController
        get() = controller
        set(controller) {
            this.controller = controller
        }

    @FXML
    private fun onBtnProcessar() {
        if (btnProcessar.accessibleText.equals("PROCESSANDO", true) && mangas != null) {
            mangas!!.setDesativar(true)
            return
        }
        btnProcessar.accessibleText = "PROCESSANDO"
        btnProcessar.text = "Pausar"
        btnCarregar.isDisable = true
        if (mangas == null) mangas = ProcessarMangas(this)
        treeBases.isDisable = true
        MenuPrincipalController.controller.getLblLog().text = "Iniciando o processamento dos mangas..."
        when (cbLinguagem.selectionModel.selectedItem) {
            Language.ENGLISH -> mangas!!.processarTabelasIngles(TABELAS)
            Language.JAPANESE -> mangas!!.processarTabelasJapones(TABELAS)
            else -> {}
        }
    }

    @FXML
    private fun onBtnCarregar() = carregar()

    @FXML
    private fun onBtnMarcarTodos() {
        marcarTodosFilhos(treeBases.root, ckbMarcarTodos.isSelected)
        treeBases.refresh()
    }

    @FXML
    private fun onBtnTransferir() {
        transferir()
    }

    fun habilitar() {
        treeBases.isDisable = false
        MenuPrincipalController.controller.getLblLog().text = ""
        btnProcessar.accessibleText = "PROCESSAR"
        btnProcessar.text = "Processar"
        btnCarregar.isDisable = false
        TaskbarProgressbar.stopProgress(Run.getPrimaryStage())
        barraProgressoVolumes.progress = 0.0
        barraProgressoCapitulos.progress = 0.0
        barraProgressoPaginas.progress = 0.0
    }

    val root: AnchorPane get() = apRoot

    private var TABELA_ORIGEM: String = ""
    private var TABELA_DESTINO: String = ""
    private var I: Long = 0
    private var error: String = ""
    private fun transferir() {
        if (txtBaseOrigem.text.trim().equals(txtBaseDestino.text.trim(), true)) {
            AlertasPopup.AvisoModal(controller.stackPane, controller.root, mutableListOf(), "Aviso", "Favor informar outra base de destino.")
            return
        }

        val progress = MenuPrincipalController.controller.criaBarraProgresso()
        progress!!.titulo.text = "Mangas - Transferencia"
        progress.log.text = "Transferindo dados...."

        btnTransferir.isDisable = true
        TABELA_ORIGEM = txtBaseOrigem.text.trim()
        TABELA_DESTINO = txtBaseDestino.text.trim()
        if (TaskbarProgressbar.isSupported()) TaskbarProgressbar.showIndeterminateProgress(Run.getPrimaryStage())
        val transferir: Task<Void> = object : Task<Void>() {
            @Override
            @Throws(Exception::class)
            override fun call(): Void? {
                try {
                    error = ""
                    updateMessage("Carregando dados....")
                    val tabelas: MutableList<String> = mutableListOf()
                    var isFull = false
                    if (TABELA_ORIGEM.contains(".*")) {
                        isFull = true
                        TABELA_ORIGEM = TABELA_ORIGEM.replace(".*", "")
                        TABELA_ORIGEM += "."
                        TABELA_ORIGEM = TABELA_ORIGEM.substring(0, TABELA_ORIGEM.lastIndexOf(".")).replace(".", "")
                        tabelas.addAll(service.getTabelasTransferir(TABELA_ORIGEM, "*"))
                    } else if (TABELA_ORIGEM.contains(".")) {
                        tabelas.add(TABELA_ORIGEM.substring(TABELA_ORIGEM.indexOf(".")))
                        TABELA_ORIGEM = TABELA_ORIGEM.substring(0, TABELA_ORIGEM.indexOf(".")) + "."
                    } else {
                        tabelas.add(TABELA_ORIGEM)
                        TABELA_ORIGEM = ""
                    }
                    for (tabela in tabelas) {
                        val lista: List<MangaVolume> = service.selectDadosTransferir(TABELA_ORIGEM, tabela)
                        if (ckbCriarBase.isSelected) {
                            updateMessage("Criando a tabela....")
                            service.createTabela(tabela)
                        }
                        updateMessage("Transferindo dados....")
                        I = 0
                        for (volume in lista) {
                            updateMessage("Transferindo dados.... " + volume.manga)
                            I++
                            if (isFull)
                                service.insertDadosTransferir(tabela, volume)
                            else
                                service.insertDadosTransferir(TABELA_DESTINO, volume)
                            updateProgress(I, lista.size.toLong())
                            Platform.runLater {
                                if (TaskbarProgressbar.isSupported())
                                    TaskbarProgressbar.showCustomProgress(Run.getPrimaryStage(), I, lista.size.toLong(), Type.NORMAL)
                            }
                        }
                    }
                } catch (e: SQLException) {
                    LOGGER.error(e.message, e)
                    error = e.message!!
                } catch (e: Error) {
                    LOGGER.error(e.message, e)
                    error = e.message!!
                }
                return null
            }

            @Override
            override fun succeeded() {
                Platform.runLater {
                    progress.barraProgresso.progressProperty().unbind()
                    progress.log.textProperty().unbind()
                    MenuPrincipalController.controller.destroiBarraProgresso(progress, "")
                    btnTransferir.isDisable = false
                    barraProgressoVolumes.progress = 0.0
                    TaskbarProgressbar.stopProgress(Run.getPrimaryStage())
                    if (error.isNotEmpty())
                        AlertasPopup.ErroModal(controller.stackPane, controller.root, mutableListOf(), "Erro", error)
                    else
                        AlertasPopup.AvisoModal(controller.stackPane, controller.root, mutableListOf(), "Aviso", "Transferência concluida.")
                }
            }
        }
        progress.barraProgresso.progressProperty().bind(transferir.progressProperty())
        progress.log.textProperty().bind(transferir.messageProperty())
        val t = Thread(transferir)
        t.start()
    }

    private var PROCESSADOS: Boolean? = null
    private var BASE: String? = null
    private var MANGA: String? = null
    private var LINGUAGEM: Language? = null
    private var DADOS: TreeItem<Manga>? = null
    private fun carregar() {
        MenuPrincipalController.controller.getLblLog().text = "Carregando dados dos mangas..."
        btnCarregar.isDisable = true
        btnProcessar.isDisable = true
        treeBases.isDisable = true
        PROCESSADOS = ckbProcessados.isSelected
        BASE = if (cbBase.value != null) cbBase.value.trim() else ""
        MANGA = txtManga.text.trim()
        LINGUAGEM = cbLinguagem.selectionModel.selectedItem
        barraProgressoCapitulos.progress = -1.0
        barraProgressoVolumes.progress = -1.0
        if (TaskbarProgressbar.isSupported()) TaskbarProgressbar.showIndeterminateProgress(Run.getPrimaryStage())

        // Criacao da thread para que esteja validando a conexao e nao trave a tela.
        val carregaItens: Task<Void> = object : Task<Void>() {
            @Override
            @Throws(Exception::class)
            override fun call(): Void? {
                try {
                    service = MangaServices()
                    TABELAS = FXCollections.observableArrayList(service.selectTabelas(!PROCESSADOS!!, false, BASE!!, LINGUAGEM!!, MANGA!!))
                    DADOS = treeData
                } catch (e: Exception) {
                    LOGGER.error(e.message, e)
                }
                return null
            }

            @Override
            override fun succeeded() {
                Platform.runLater {
                    treeBases.setRoot(DADOS)
                    MenuPrincipalController.controller.getLblLog().text = ""
                    ckbMarcarTodos.isSelected = true
                    btnCarregar.isDisable = false
                    btnProcessar.isDisable = false
                    treeBases.isDisable = false
                    barraProgressoCapitulos.progress = 0.0
                    barraProgressoVolumes.progress = 0.0
                    TaskbarProgressbar.stopProgress(Run.getPrimaryStage())
                }
            }
        }
        val t = Thread(carregaItens)
        t.start()
    }

    // Implementa um nivel por tipo
    private val treeData: TreeItem<Manga>
        get() {
            val itmRoot: TreeItem<Manga> = TreeItem(Manga("...", ""))
            for (tabela in TABELAS) {
                tabela.manga = "..."
                val itmTabela: TreeItem<Manga> = TreeItem(tabela)
                var itmManga: TreeItem<Manga>? = null
                var volumeAnterior = ""
                for (volume in tabela.volumes) {
                    // Implementa um nivel por tipo
                    if (!volume.manga.equals(volumeAnterior, true) || itmManga == null) {
                        volumeAnterior = volume.manga
                        volume.base = tabela.base
                        itmManga = TreeItem(Manga(tabela.base, volume.manga, "..."))
                        itmTabela.children.add(itmManga)
                        itmTabela.isExpanded = true
                    }
                    volume.base = tabela.base
                    val itmVolume: TreeItem<Manga> = TreeItem(volume)
                    for (capitulo in volume.capitulos) {
                        capitulo.base = tabela.base
                        capitulo.nomePagina = "..."
                        val itmCapitulo: TreeItem<Manga> = TreeItem(capitulo)
                        for (pagina in capitulo.paginas) {
                            pagina.addOutrasInformacoes(tabela.base, volume.manga, volume.volume, capitulo.capitulo, volume.lingua)
                            itmCapitulo.children.add(TreeItem(pagina))
                        }
                        itmVolume.children.add(itmCapitulo)
                    }
                    itmManga.children.add(itmVolume)
                }
                itmRoot.children.add(itmTabela)
                itmRoot.isExpanded = true
            }
            return itmRoot
        }

    private fun marcarTodosFilhos(treeItem: TreeItem<Manga>, newValue: Boolean) {
        treeItem.value.isProcessar = newValue
        treeItem.children.forEach { treeItemNivel2 -> marcarTodosFilhos(treeItemNivel2, newValue) }
    }

    private fun ativaTodosPai(treeItem: TreeItem<Manga>, newValue: Boolean) {
        if (treeItem.parent != null) {
            treeItem.parent.value.isProcessar = newValue
            ativaTodosPai(treeItem.parent, newValue)
        }
    }

    private fun editaColunas() {
        // ==== (CHECK-BOX) ===
        treecMacado.setCellValueFactory { param ->
            val treeItem: TreeItem<Manga> = param.value
            val item: Manga = treeItem.value
            val booleanProp = SimpleBooleanProperty(item.isProcessar)
            booleanProp.addListener { _, _, newValue ->
                item.isProcessar = newValue
                marcarTodosFilhos(treeItem, newValue)
                if (newValue) // Somente ativa caso seja true, pois ao menos um nó precisa estar ativo
                    ativaTodosPai(treeItem, newValue)
                treeBases.refresh()
            }
            booleanProp
        }
        treecMacado.setCellFactory {
            val cell: CheckBoxTreeTableCellCustom<Manga, Boolean> = CheckBoxTreeTableCellCustom()
            cell.alignment = Pos.CENTER
            cell
        }
    }

    private fun linkaCelulas() {
        treecMacado.cellValueFactory = TreeItemPropertyValueFactory("processar")
        treecBase.setCellValueFactory(TreeItemPropertyValueFactory("base"))
        treecManga.setCellValueFactory(TreeItemPropertyValueFactory("manga"))
        treecVolume.setCellValueFactory(TreeItemPropertyValueFactory("volume"))
        treecCapitulo.setCellValueFactory(TreeItemPropertyValueFactory("capitulo"))
        treecPagina.setCellValueFactory(TreeItemPropertyValueFactory("pagina"))
        treecNomePagina.setCellValueFactory(TreeItemPropertyValueFactory("nomePagina"))
        treeBases.isShowRoot = false
        editaColunas()
    }

    private val robot: Robot = Robot()
    override fun initialize(arg0: URL, arg1: ResourceBundle) {
        cbLinguagem.items.addAll(Language.JAPANESE, Language.ENGLISH)
        cbLinguagem.selectionModel.selectFirst()

        try {
            cbBase.items.setAll(service.tabelas)
        } catch (e: Exception) {
            LOGGER.error(e.message, e)
        }

        val autoCompletePopup: JFXAutoCompletePopup<String> = JFXAutoCompletePopup()
        autoCompletePopup.suggestions.addAll(cbBase.items)
        autoCompletePopup.setSelectionHandler { event -> cbBase.setValue(event.getObject()) }
        cbBase.editor.textProperty().addListener { _, _, _ ->
            autoCompletePopup.filter { item -> item.lowercase(Locale.getDefault()).contains(cbBase.editor.text.lowercase(Locale.getDefault())) }
            if (autoCompletePopup.filteredSuggestions.isEmpty() || cbBase.showingProperty().get() || cbBase.editor.text.isEmpty())
                autoCompletePopup.hide()
            else
                autoCompletePopup.show(cbBase.editor)
        }
        cbBase.setOnKeyPressed { ke -> if (ke.code.equals(KeyCode.ENTER)) robot.keyPress(KeyCode.TAB) }
        txtManga.setOnKeyPressed { ke -> if (ke.code.equals(KeyCode.ENTER)) robot.keyPress(KeyCode.TAB) }
        linkaCelulas()
    }

    companion object {
        val fxmlLocate: URL get() = MangasProcessarController::class.java.getResource("/view/mangas/MangaProcessar.fxml") as URL
    }
}